File: src/main/java/net/ilkinulas/redis/RedisQueue.java
package net.ilkinulas.redis;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import redis.clients.jedis.*;

import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.List;

class RedisQueue implements Queue, AutoCloseable {
    private static final Logger logger = LoggerFactory.getLogger(RedisQueue.class);
    private final String host;
    private final int port;
    private final String queueName;

    private JedisPool pool;

    RedisQueue(String host, int port, String name) {
        this.host = host;
        this.port = port;
        this.queueName = name;
    }

    void connect() {
        pool = new JedisPool(new JedisPoolConfig(), host, port);
        logger.info("RedisQueue is ready. Connected to {}:{}", host, port);
    }

    @Override
    public void close() {
        pool.close();
    }

    @Override
    public String poll() {
        try (Jedis redis = pool.getResource()) {
            return redis.lpop(queueName);
        }
    }
    
    @Override
    public long size() {
        try (Jedis redis = pool.getResource()) {
            return redis.llen(queueName);
        }
    }

    @Override
    public long add(String s) {
        try (Jedis redis = pool.getResource()) {
            return redis.rpush(queueName, s);
        }
    }

    @Override
    public long add(List<String> l) {
        try (Jedis redis = pool.getResource()) {
            return redis.rpush(queueName, l.toArray(new String[l.size()]));
        }
    }

    @Override
    public List<String> poll(int count) {
        try (Jedis redis = pool.getResource()) {
            Transaction tx = redis.multi();
            Response<List<String>> response = tx.lrange(queueName, 0, count - 1);
            tx.ltrim(queueName, count, -1);
            tx.exec();
            return response.get();
        }
    }

    @Override
    public void clear() {
        try (Jedis redis = pool.getResource()) {
            redis.del(queueName);
        }
    }

    public List<String> poplua(long count) throws Exception {
        String script = readFromClassPath("lpopn.lua");
        try (Jedis redis = pool.getResource()) {
            Object result = redis.eval(script, 1, queueName, String.valueOf(count));
            if (result == null) {
                return Collections.emptyList();
            }
            return (List<String>) result;
        }
    }

    private String readFromClassPath(final String fileName) throws URISyntaxException, IOException {
        return new String(Files.readAllBytes(
                Paths.get(getClass().getClassLoader()
                        .getResource(fileName)
                        .toURI())));
    }
}

File: src/main/kotlin/net/ilkinulas/redis/RedisQueue.kt
package net.ilkinulas.redis

import org.slf4j.LoggerFactory
import redis.clients.jedis.JedisPool
import redis.clients.jedis.JedisPoolConfig

interface Queue {
    fun poll(): String?
    fun poll(count: Long): List<String>
    fun size(): Long
    fun add(s: String): Long
    fun add(l: List<String>): Long
    fun clear()
}

private val logger = LoggerFactory.getLogger(RedisQueue::class.java)

class RedisQueue(
        host: String = "localhost",
        port: Int = 6379,
        private val queue: String) : Queue, AutoCloseable {

    private var pool: JedisPool

    init {
        val poolConfig = JedisPoolConfig()
        pool = JedisPool(poolConfig, host, port)
        logger.info("RedisQueue is ready. Connected to $host:$port")
    }

    override fun poll(): String? {
        pool.resource.use {
            return it.lpop(queue)
        }
    }

    override fun poll(count: Long): List<String> {
        pool.resource.use {
            val tx = it.multi()
            val response = tx.lrange(queue, 0, count - 1)
            tx.ltrim(queue, count, -1)
            tx.exec()
            return response.get()
        }
    }

    override fun size() = pool.resource.use { it.llen(queue) }

    override fun add(s: String): Long {
        pool.resource.use {
            return it.rpush(queue, s)
        }
    }

    override fun add(l: List<String>): Long {
        pool.resource.use {
            return it.rpush(queue, *l.toTypedArray())
        }
    }

    override fun clear() {
        pool.resource.use {
            it.del(queue)
        }
    }

    override fun close() {
        pool.close()
    }

}